src/utils/format_json/open_telemetry.c

```diff
@@ -0,0 +1,278 @@
+  // CHECK(json_add_string(g, "1"));
+
+  if (fam->help != NULL) {
+    CHECK(json_add_string(g, "description"));
+    CHECK(json_add_string(g, fam->help));
+  }
+
+  switch (fam->type) {
+  case METRIC_TYPE_COUNTER:
+    CHECK(json_add_string(g, "sum"));
+    CHECK(sum(g, fam));
+    break;
+  case METRIC_TYPE_GAUGE:
+    CHECK(json_add_string(g, "gauge"));
+    CHECK(gauge(g, fam));
+    break;
+  case METRIC_TYPE_UNTYPED:
+    // TODO
+    assert(0);
+  }
+
+  CHECK(yajl_gen_map_close(g)); /* END Metric */
+  return 0;
+}
+
+static int instrumentation_scope(yajl_gen g) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN InstrumentationScope */
+
+  CHECK(json_add_string(g, "name"));
+  CHECK(json_add_string(g, PACKAGE_NAME));
+
+  CHECK(json_add_string(g, "version"));
+  CHECK(json_add_string(g, PACKAGE_VERSION));
+
+  CHECK(yajl_gen_map_close(g)); /* END InstrumentationScope */
+  return 0;
+}
+
+static int scope_metrics(yajl_gen g, metric_family_t const *fam) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN ScopeMetrics */
+
+  CHECK(json_add_string(g, "scope"));
+  CHECK(instrumentation_scope(g));
+
+  CHECK(json_add_string(g, "metrics"));
+  CHECK(yajl_gen_array_open(g));
+  CHECK(metric(g, fam));
+  CHECK(yajl_gen_array_close(g));
+
+  CHECK(yajl_gen_map_close(g)); /* END ScopeMetrics */
+  return 0;
+}
+
+static int resource(yajl_gen g, label_set_t res) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN Resource */
+
+  CHECK(json_add_string(g, "attributes"));
+  CHECK(yajl_gen_array_open(g));
+  for (size_t i = 0; i < res.num; i++) {
+    CHECK(key_value(g, res.ptr[i]));
+  }
+  CHECK(yajl_gen_array_close(g));
+
+  CHECK(yajl_gen_map_close(g)); /* END Resource */
+  return 0;
+}
+
+static int add_resource_metric(yajl_gen g, metric_family_t const *fam) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN ResourceMetrics */
+
+  if (fam->resource.num > 0) {
+    CHECK(json_add_string(g, "resource"));
+    CHECK(resource(g, fam->resource));
+  }
+
+  CHECK(json_add_string(g, "scopeMetrics"));
+  CHECK(yajl_gen_array_open(g));
+  CHECK(scope_metrics(g, fam));
+  CHECK(yajl_gen_array_close(g));
+
+  CHECK(yajl_gen_map_close(g)); /* END ResourceMetrics */
+  return 0;
+}
+
+int format_json_open_telemetry(strbuf_t *buf, metric_family_t const **families,
+                               size_t families_num) {
+  if (buf->pos != 0) {
+    ERROR("format_json_open_telemetry: buffer is not empty.");
+    return EINVAL;
+  }
+
+  yajl_gen g = yajl_gen_alloc(NULL);
+  if (g == NULL) {
+    ERROR("format_json_open_telemetry: yajl_gen_alloc() failed.");
+    return ENOMEM;
+  }
+#if COLLECT_DEBUG
+  yajl_gen_config(g, yajl_gen_validate_utf8, 1);
+#endif
+
+  // TODO
+  CHECK(yajl_gen_map_open(g)); /* BEGIN ExportMetricsServiceRequest */
+  CHECK(json_add_string(g, "resourceMetrics"));
+  CHECK(yajl_gen_array_open(g));
+
+  unsigned char const *out = NULL;
+  for (size_t i = 0; i < families_num; i++) {
+    metric_family_t const *fam = families[i];
+    add_resource_metric(g, fam);
```

Comment:
As the error message is always same, error value is always propagated up, and this is the only place where those other (static) functions are called, it would be best to drop logging from `CHECK()`, and do logging for the error values only in this function.

Reply:
I just pushed two changes:

* Add specialized logging to `json_add_string()` so it includes the string that it failed to add to the buffer.
* Modify the macro to only log an error if the function name begins with `yajl_gen_`.

I think this should address the concerns you have raised.